### PR TITLE
Implement Matchismo game

### PR DESCRIPTION
**Pull request by Shiaulis**
Lecture assignment text can be found here:
https://drive.google.com/drive/u/1/folders/1yoBo_pjdzNq6nx2_0d42x6TQ8gKj0PZp

Implemented:
- new game button
- game mode segmented control
- ability to match 3 cards
- minor style and code structure improvements

### CODE REVIEW

File: Matchismo/CardGameViewController.m [hunk 1]
```diff
@@ -14,41 +14,107 @@ @interface CardGameViewController ()
...
 
-- (CardMatchingGame *)game
-{
-    if (_game == nil) {
-        _game = [[CardMatchingGame alloc] initWithCardCount:self.cardButtons.count usingDeck:[self createDeck]];
-    }
+#pragma - View Controller Life Cycle
 
-    return _game;
+- (void)viewDidLoad
```
Comment (khaletska) [inline Matchismo/CardGameViewController.m:31]: this is forbidden magic here

File: Matchismo/CardGameViewController.m [hunk 2]
```diff
@@ -14,41 +14,107 @@ @interface CardGameViewController ()
...
+
 - (void)updateUI
+{
+    [self updateCards];
+    [self updateGameModeSwitch];
+    [self updateNewGameButton];
+    [self updateScoreCounter];
+    [self updateStatusLabel];
+
+//    NSLog(@"UI Updated");
```
Comment (khaletska) [inline Matchismo/CardGameViewController.m]: logs can be removed 😊

File: Matchismo/CardGameViewController.m [hunk 3]
```diff
@@ -57,10 +123,28 @@ - (void)updateUI
...
+{
+    self.gameModeSwitch.enabled = ![self didGameStart];
+}
+
+- (void)updateNewGameButton
+{
+    self.startNewGameButton.hidden = ![self didGameStart];
+}
+
+- (void)updateScoreCounter
```
Comment (khaletska) [inline Matchismo/CardGameViewController.m]:
```suggestion
- (void)updateScoreLabel
```

File: Matchismo/CardGameViewController.m [hunk 4]
```diff
@@ -14,41 +14,107 @@ @interface CardGameViewController ()
...
+
+- (void)discardCurrentGame
+{
+    self.game = nil;
+    [self updateUI];
+}
+
+- (void)startNewGameIfNeeded
+{
+    if (self.game != nil) {
```
Comment (khaletska) [inline Matchismo/CardGameViewController.m]: you have a method for it
```suggestion
    if ([self didGameStart]) {
```

File: Matchismo/Model/CardMatchingGame.m [hunk 5]
```diff
@@ -63,30 +73,91 @@ - (void)chooseCardAtIndex:(NSUInteger)index
...
-                        self.score -= MISMATCH_PENALTY;
-                        otherCard.chosen = NO;
-                    }
-                    break;
-                }
-            }
-            self.score -= COST_TO_CHOOSE;
-            card.chosen = YES;
+            NSArray *mismatchedCards = [chosenCards arrayByAddingObject:card];
+            NSInteger receivedPenalty = self.mismatchPenalty;
```
Comment (khaletska) [inline Matchismo/Model/CardMatchingGame.m]: why do we need this variable?

File: Matchismo/Model/CardMatchingGame.m [hunk 6]
```diff
@@ -63,30 +73,91 @@ - (void)chooseCardAtIndex:(NSUInteger)index
...
+            self.score -= receivedPenalty;
+            [self deselectCards:chosenCards];
+            self.state = [NSString stringWithFormat:@"%@ don't match! %ld points penalty!", [CardMatchingGame makeDescriptionForCards:mismatchedCards], receivedPenalty];
+        }
+    }
+    else {
+        NSArray *selectedCards = [chosenCards arrayByAddingObject:card];
```
Comment (khaletska) [inline Matchismo/Model/CardMatchingGame.m:112]: it is always gonna be just one card that was last selected, isn't it?

File: Matchismo/Model/CardMatchingGame.m [hunk 7]
```diff
@@ -98,4 +169,35 @@ - (Card *)cardAtIndex:(NSUInteger)index
...
+    for (Card *card in cards) {
+        if (description.length == 0) {
+            [description appendString:card.contents];
+        }
+        else {
+            [description appendFormat:@", %@", card.contents];
+        }
+    }
+
+    return [description copy];
```
Comment (khaletska) [inline Matchismo/Model/CardMatchingGame.m:199]: why copy?

File: Matchismo/Model/PlayingCard.m [hunk 8]
```diff
@@ -36,20 +36,38 @@ - (void)setRank:(NSUInteger)rank
 
 #pragma - Overriden methods -
```
Comment (khaletska) [inline Matchismo/Model/PlayingCard.m:37]:
```suggestion
#pragma - Overwritten methods -
```

File: Matchismo/Model/PlayingCard.m [hunk 9]
```diff
@@ -36,20 +36,38 @@ - (void)setRank:(NSUInteger)rank
...
-    if (otherCards.count == 1) {
-        PlayingCard *otherCard = (PlayingCard *)otherCards.firstObject;
-        if ([self.suit isEqualToString:otherCard.suit]) {
-            score = 1;
+    // According to requirements it is fine to assume to have not more than 3 cards
+    PlayingCard *cardOne = self;
+    PlayingCard *cardTwo = otherCards.firstObject;
+    PlayingCard *cardThree = otherCards.count > 1 ? otherCards[1] : nil;
+
+    if ([cardOne.suit isEqualToString:cardTwo.suit] || [cardOne.suit isEqualToString:cardThree.suit] || [cardTwo.suit isEqualToString:cardThree.suit]) {
```
Comment (khaletska) [inline Matchismo/Model/PlayingCard.m]:
```suggestion
    if ([cardOne.suit isEqualToString:cardTwo.suit] || 
        [cardOne.suit isEqualToString:cardThree.suit]
        [cardTwo.suit isEqualToString:cardThree.suit]) {
```

Comment (khaletska) [reply Matchismo/Model/PlayingCard.m]: for sake of readability 😊

**Review by khaletska — CHANGES_REQUESTED** [review]
Something seems to be super wrong here

image

File: Matchismo/Model/PlayingCard.m (hunk 8 above)
Comment (Shiaulis) [reply Matchismo/Model/PlayingCard.m:37]: image

File: Matchismo/Model/CardMatchingGame.m (hunk 5 above)
Comment (Shiaulis) [reply Matchismo/Model/CardMatchingGame.m]: to pass it to game state string

Comment (khaletska) [reply Matchismo/Model/CardMatchingGame.m]: can you just pass self.mismatchPenalty?

File: Matchismo/Model/CardMatchingGame.m (hunk 6 above)
Comment (Shiaulis) [reply Matchismo/Model/CardMatchingGame.m:112]: I assume yes, why?

Comment (khaletska) [reply Matchismo/Model/CardMatchingGame.m:112]: so you don't need another array, you can just use card

**Review by khaletska — APPROVED** [review]
Great job 😊

**Pull request merged**